### PR TITLE
Dispatcher refactor

### LINKED ISSUE #21: Dispatcher re-architecture
In the current design, each subscription gets its own dispatcher deployment.  This allows for maximum isolation and independence of scale of each subscription.  However, it comes with a high resource cost, especially for installations with costly sidecars on each pod.  An option for redesign is to collapse dispatchers down to a single deployment per channel.  This potentially strikes a good balance between isolation of different event streams (high volume/low volume, sensitive/non-sensitive, etc), but increases resource efficiency when many subscriptions exist for the same channel.  These deployments would support horizontal scale with a replica count up to the number of partitions of the underlying kafka topic.

An understanding of resource cost of this consolidation should be gained to evaluate if additional changes are required on top. Each consumer of a consumer group with librdkafka carries a significant cost on its own (caching, polling, etc).  Measurements of memory and cpu usage will help decide how to manage number of consumers per consumer group in a single pod.

### PR DESCRIPTION
**Pull request by phamilton**

### CODE REVIEW

File: components/dispatcher/internal/client/client.go [hunk 1]
```diff
@@ -37,10 +37,9 @@ type retriableCloudEventClient struct {
...
 var _ RetriableClient = &retriableCloudEventClient{}
 
-func NewRetriableCloudEventClient(uri string, exponentialBackoff bool, initialRetryInterval int64, maxRetryTime int64) retriableCloudEventClient {
+func NewRetriableCloudEventClient(exponentialBackoff bool, initialRetryInterval int64, maxRetryTime int64) retriableCloudEventClient {
 	transport, err := cloudevents.NewHTTPTransport(
-		cloudevents.WithTarget(uri),
-		cloudevents.WithEncoding(cloudevents.HTTPBinaryV03),
+		cloudevents.WithBinaryEncoding(),
 	)
 	transport.Client = httpClient
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/client/client.go:44]: transport could be nil here - should move to after error handling?

File: components/dispatcher/internal/client/client.go [hunk 2]
```diff
@@ -53,27 +52,26 @@ func NewRetriableCloudEventClient(uri string, exponentialBackoff bool, initialRe
 		panic("unable to create cloudevent client: " + err.Error())
 	}
 
-	return retriableCloudEventClient{uri: uri, exponentialBackoff: exponentialBackoff, initialRetryInterval: initialRetryInterval, maxRetryTime: maxRetryTime, cloudEventClient: ceClient}
+	return retriableCloudEventClient{exponentialBackoff: exponentialBackoff, initialRetryInterval: initialRetryInterval, maxRetryTime: maxRetryTime, cloudEventClient: ceClient}
 }
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/client/client.go:61]: expand (multi-line) this struct for readability?

File: components/dispatcher/internal/dispatcher/dispatcher.go [hunk 3]
```diff
@@ -166,29 +158,61 @@ func (d *Dispatcher) handleKafkaMessages(consumer kafkaconsumer.ConsumerInterfac
 }
 
 // Store Updated Offsets For The Partition If Consumer Still Has It Assigned
-func (d *Dispatcher) updateOffsets(logger *zap.Logger, consumer kafkaconsumer.ConsumerInterface, message *kafka.Message) {
+func (d *Dispatcher) updateOffsets(consumer kafkaconsumer.ConsumerInterface, message *kafka.Message) {
 	// Store The Updated Offsets
 	offsets := []kafka.TopicPartition{message.TopicPartition}
 	offsets[0].Offset++
 	consumer.StoreOffsets(offsets)
 }
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher.go:173]: should handle error from StoreOffsets() ?

File: components/dispatcher/internal/dispatcher/dispatcher.go [hunk 4]
```diff
@@ -166,29 +158,61 @@ func (d *Dispatcher) handleKafkaMessages(consumer kafkaconsumer.ConsumerInterfac
...
-			d.offsetMessages[index][partition.Partition] = partition.Offset
+			consumerOffset.offsets[partition.Partition] = partition.Offset
 		}
 	}
-	d.lastOffsetCommit[index] = time.Now()
+	consumerOffset.lastOffsetCommit = time.Now()
+}
+
+func (d *Dispatcher) UpdateSubscriptions(subscriptions []Subscription) (map[Subscription]error, error) {
+
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher.go:194]: why does this return an error when it never returns one ?

File: components/dispatcher/internal/dispatcher/dispatcher_test.go [hunk 5]
```diff
@@ -33,16 +34,15 @@ const (
...
-	numberOfCalls      = 0
-	numberOfCallsMutex = &sync.Mutex{}
-	testValue          = map[string]string{"test": "value"}
+	topic            = testTopic
+	testError        = kafka.NewError(1, "sample error", false)
+	testNotification = kafka.RevokedPartitions{}
+	testValue        = map[string]string{"test": "value"}
 )
 
 // Setup A Test Logger (Ignore Unused Warning - This Updates The log.Logger Reference!)
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher_test.go]: logger on line below is never used - can be removed

File: components/dispatcher/internal/dispatcher/dispatcher_test.go [hunk 6]
```diff
@@ -93,18 +101,46 @@ func TestDispatcher(t *testing.T) {
...
+		},
+	})
 
 	// Verify Consumers Started
-	verifyConsumersCount(t, testDispatcher.consumers)
+	verifyConsumersCount(t, testDispatcher.consumers, 1)
+
+	// Remove All Consumers
+	testDispatcher.UpdateSubscriptions(make([]Subscription, 0))
+
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher_test.go:125]: all these updateSubscriptions() calls should have an assert that returned err is nil?  Unless we remove the error response from the function as called for in another PR comment?

File: components/dispatcher/internal/dispatcher/dispatcher.go [hunk 7]
```diff
@@ -166,29 +158,61 @@ func (d *Dispatcher) handleKafkaMessages(consumer kafkaconsumer.ConsumerInterfac
...
 	if err != nil {
 		// Don't Log Error If There Weren't Any Offsets To Commit
 		if err.(kafka.Error).Code() != kafka.ErrNoOffset {
 			logger.Error("Error Committing Offsets", zap.Error(err))
 		}
 	} else {
-		logger.Info("Committed Partitions", zap.Any("partitions", partitions))
-		d.offsetMessages[index] = make(map[int32]kafka.Offset)
+		//logger.Debug("Committed Partitions", zap.Any("partitions", partitions))
+		consumerOffset.offsets = make(map[int32]kafka.Offset)
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher.go:185]: enable the debug log statement so we have a way to see that an event was processed/committed

File: components/dispatcher/internal/dispatcher/dispatcher_test.go [hunk 8]
```diff
@@ -187,56 +227,54 @@ func verifyDispatcher(t *testing.T,
...
+func verifyConsumerCommits(t *testing.T, consumers map[Subscription]*ConsumerOffset, expectedCount int) {
 	assert.NotNil(t, consumers)
-	assert.Len(t, consumers, testConcurrency)
+	assert.Len(t, consumers, expectedCount)
 	for _, consumer := range consumers {
-		mockConsumer, ok := consumer.(*MockConsumer)
-		assert.True(t, mockConsumer.closed)
+		mockConsumer, ok := consumer.consumer.(*MockConsumer)
+		//assert.True(t, mockConsumer.closed)
 		assert.True(t, ok)
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/dispatcher/dispatcher_test.go:281]: remove unused assert?

File: components/dispatcher/internal/testing/factory.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+			addTo(dynamicScheme)
+		}
+
+		eventRecorder := record.NewFakeRecorder(maxEventBufferSize)
+		statsReporter := &FakeStatsReporter{}
+
+		// Set up our Controller from the fakes.
+		// Eventing
+		//	ncInformer := messagingInformerFactory.Knativekafka().V1alpha1().KafkaChannels()
+
```
Comment (travis-minke-sap) [inline components/dispatcher/internal/testing/factory.go]: remove or comment on why keeping???

File: components/dispatcher/internal/client/client.go (hunk 1 above)
Comment (phamilton) [reply components/dispatcher/internal/client/client.go:44]: Good call.

File: components/dispatcher/internal/client/client.go (hunk 2 above)
Comment (phamilton) [reply components/dispatcher/internal/client/client.go:61]: 👍

**Comment by phamilton** [comment]
Also added discussed changes to support other cloudevent versions

**Review by phamilton — COMMENTED** [review]
@travis-minke-sap your changes, LGTM

**Review by travis-minke-sap — APPROVED** [review]

**Pull request merged**